File: backend/core/src/modules/beta-sessions/beta-sessions.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { InjectRepository } from '@nestjs/typeorm';
import { createHash, randomBytes } from 'crypto';
import { In, Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { BetaAccessService } from '../beta-access/beta-access.service';
import { BetaSessionInvite, BetaSessionParticipant, BetaSession, Document, User } from '../../entities';
import { CreateBetaSessionDto } from './dto/create-beta-session.dto';
import { UpdateBetaSessionDto } from './dto/update-beta-session.dto';
import { CreateBetaSessionInviteDto } from './dto/create-beta-session-invite.dto';

@Injectable()
export class BetaSessionsService {
  constructor(
    private readonly configService: ConfigService,
    private readonly projectAccessService: ProjectAccessService,
    private readonly betaAccessService: BetaAccessService,
    @InjectRepository(BetaSession)
    private readonly sessionsRepo: Repository<BetaSession>,
    @InjectRepository(BetaSessionInvite)
    private readonly invitesRepo: Repository<BetaSessionInvite>,
    @InjectRepository(BetaSessionParticipant)
    private readonly participantsRepo: Repository<BetaSessionParticipant>,
    @InjectRepository(Document)
    private readonly documentsRepo: Repository<Document>,
    @InjectRepository(User)
    private readonly usersRepo: Repository<User>,
  ) {}

  private hashToken(raw: string) {
    return createHash('sha256').update(raw).digest('hex');
  }

  async listForUser(userId: string | undefined, projectId?: string) {
    if (!userId) return [];
    const projectIds = await this.projectAccessService.filterAccessibleProjectIds(userId, projectId);
    if (projectIds.length === 0) return [];
    return this.sessionsRepo.find({ where: { projectId: In(projectIds) } as any, order: { createdAt: 'DESC' } as any });
  }

  async findOneForUser(userId: string | undefined, id: string) {
    if (!userId) return null;
    const { session } = await this.betaAccessService.getAccess(userId, id);
    return session;
  }

  async createForUser(userId: string | undefined, dto: CreateBetaSessionDto) {
    if (!userId) throw new BadRequestException('로그인이 필요합니다.');
    if (!dto.projectId) throw new BadRequestException('projectId가 필요합니다.');
    if (!dto.title) throw new BadRequestException('title이 필요합니다.');

    await this.projectAccessService.assertProjectAccess(userId, dto.projectId);
    const created = this.sessionsRepo.create({
      projectId: dto.projectId,
      documentId: dto.documentId,
      title: dto.title,
      description: dto.description,
      status: dto.status ?? 'open',
      createdBy: userId,
    });
    return this.sessionsRepo.save(created);
  }

  async updateForUser(userId: string | undefined, id: string, dto: UpdateBetaSessionDto) {
    if (!userId) throw new BadRequestException('로그인이 필요합니다.');
    const session = await this.betaAccessService.assertManageAccess(userId, id);

    await this.sessionsRepo.update(
      { id: session.id } as any,
      {
        ...(dto.title !== undefined ? { title: dto.title } : {}),
        ...(dto.description !== undefined ? { description: dto.description } : {}),
        ...(dto.status !== undefined ? { status: dto.status } : {}),
      } as any,
    );
    return this.sessionsRepo.findOne({ where: { id: session.id } });
  }

  async removeForUser(userId: string | undefined, id: string) {
    if (!userId) return;
    const session = await this.betaAccessService.assertManageAccess(userId, id);
    await this.sessionsRepo.delete({ id: session.id } as any);
  }

  async createInviteForSession(userId: string | undefined, sessionId: string, dto: CreateBetaSessionInviteDto) {
    if (!userId) throw new BadRequestException('로그인이 필요합니다.');
    await this.betaAccessService.assertManageAccess(userId, sessionId);

    const rawToken = randomBytes(32).toString('hex');
    const tokenHash = this.hashToken(rawToken);

    const hours = dto.expiresInHours ?? 24 * 7;
    const expiresAt = hours ? new Date(Date.now() + hours * 60 * 60 * 1000) : undefined;

    const invite = await this.invitesRepo.save(
      this.invitesRepo.create({
        sessionId,
        tokenHash,
        expiresAt,
        maxUses: dto.maxUses,
        createdBy: userId,
      }),
    );

    const baseUrl = this.configService.get<string>('app.baseUrl') ?? 'http://localhost:3100';
    const inviteUrl = `${baseUrl.replace(/\/$/, '')}/beta-invite?token=${encodeURIComponent(rawToken)}`;

    return {
      success: true,
      inviteId: invite.id,
      token: rawToken,
      inviteUrl,
      expiresAt: invite.expiresAt,
      maxUses: invite.maxUses,
    };
  }

  async getInviteInfo(userId: string | undefined, token: string) {
    if (!userId) throw new BadRequestException('로그인이 필요합니다.');
    if (!token) throw new BadRequestException('token이 필요합니다.');

    const tokenHash = this.hashToken(token);
    const invite = await this.invitesRepo.findOne({ where: { tokenHash } as any });
    if (!invite) throw new BadRequestException('초대장을 찾을 수 없습니다.');
    if (invite.expiresAt && invite.expiresAt.getTime() < Date.now()) {
      throw new BadRequestException('초대장이 만료되었습니다.');
    }
    if (invite.maxUses !== undefined && invite.uses >= invite.maxUses) {
      throw new BadRequestException('초대장 사용 횟수를 초과했습니다.');
    }

    const session = await this.sessionsRepo.findOne({ where: { id: invite.sessionId } });
    if (!session) throw new BadRequestException('베타 세션을 찾을 수 없습니다.');

    return {
      session: {
        id: session.id,
        projectId: session.projectId,
        documentId: session.documentId,
        title: session.title,
        description: session.description,
        status: session.status,
      },
      invite: {
        id: invite.id,
        expiresAt: invite.expiresAt,
        maxUses: invite.maxUses,
        uses: invite.uses,
      },
    };
  }

  async joinByInvite(userId: string | undefined, token: string) {
    if (!userId) throw new BadRequestException('로그인이 필요합니다.');
    if (!token) throw new BadRequestException('token이 필요합니다.');

    const tokenHash = this.hashToken(token);
    const invite = await this.invitesRepo.findOne({ where: { tokenHash } as any });
    if (!invite) throw new BadRequestException('초대장을 찾을 수 없습니다.');
    if (invite.expiresAt && invite.expiresAt.getTime() < Date.now()) {
      throw new BadRequestException('초대장이 만료되었습니다.');
    }
    if (invite.maxUses !== undefined && invite.uses >= invite.maxUses) {
      throw new BadRequestException('초대장 사용 횟수를 초과했습니다.');
    }

    const session = await this.sessionsRepo.findOne({ where: { id: invite.sessionId } });
    if (!session) throw new BadRequestException('베타 세션을 찾을 수 없습니다.');

    const existing = await this.participantsRepo.findOne({ where: { sessionId: session.id, userId } as any });
    if (!existing) {
      const user = await this.usersRepo.findOne({ where: { id: userId } as any, select: { name: true } as any });
      await this.participantsRepo.save(
        this.participantsRepo.create({
          sessionId: session.id,
          userId,
          status: 'joined',
          displayName: user?.name,
        }),
      );
      await this.invitesRepo.update({ id: invite.id } as any, { uses: invite.uses + 1 } as any);
    }

    return {
      success: true,
      sessionId: session.id,
    };
  }

  async listParticipantsForSession(userId: string | undefined, sessionId: string) {
    if (!userId) throw new BadRequestException('로그인이 필요합니다.');
    const session = await this.betaAccessService.assertManageAccess(userId, sessionId);

    const participants = await this.participantsRepo.find({ where: { sessionId: session.id } as any, order: { joinedAt: 'ASC' } as any });
    if (participants.length === 0) return [];

    const users = await this.usersRepo.find({
      where: { id: In(participants.map((p) => p.userId)) } as any,
      select: { id: true, name: true, avatarUrl: true } as any,
    });
    const userMap = new Map(users.map((u) => [u.id, u]));

    return participants.map((p) => ({
      ...p,
      user: userMap.get(p.userId) ?? null,
    }));
  }

  async getSessionDocumentForUser(userId: string | undefined, sessionId: string) {
    if (!userId) throw new BadRequestException('로그인이 필요합니다.');
    const { session } = await this.betaAccessService.getAccess(userId, sessionId);
    if (!session.documentId) {
      throw new BadRequestException('이 세션에는 documentId가 설정되어 있지 않습니다.');
    }

    const doc = await this.documentsRepo.findOne({
      where: { id: session.documentId } as any,
      select: { id: true, projectId: true, title: true, content: true, updatedAt: true } as any,
    });
    if (!doc) throw new BadRequestException('문서를 찾을 수 없습니다.');
    if (doc.projectId !== session.projectId) throw new BadRequestException('문서 정보가 올바르지 않습니다.');

    return doc;
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/BetaSessionsService.java
package store.pjcloud.cowrite.core.service;

import java.security.MessageDigest;
import java.security.SecureRandom;
import java.time.OffsetDateTime;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;
import store.pjcloud.cowrite.core.config.AppProperties;
import store.pjcloud.cowrite.core.entity.BetaSession;
import store.pjcloud.cowrite.core.entity.BetaSessionInvite;
import store.pjcloud.cowrite.core.entity.BetaSessionParticipant;
import store.pjcloud.cowrite.core.entity.Document;
import store.pjcloud.cowrite.core.entity.User;
import store.pjcloud.cowrite.core.repository.BetaSessionInviteRepository;
import store.pjcloud.cowrite.core.repository.BetaSessionParticipantRepository;
import store.pjcloud.cowrite.core.repository.BetaSessionRepository;
import store.pjcloud.cowrite.core.repository.DocumentRepository;
import store.pjcloud.cowrite.core.repository.UserRepository;

@Service
public class BetaSessionsService {
    private final AppProperties appProperties;
    private final ProjectAccessService projectAccessService;
    private final BetaAccessService betaAccessService;
    private final BetaSessionRepository sessionsRepo;
    private final BetaSessionInviteRepository invitesRepo;
    private final BetaSessionParticipantRepository participantsRepo;
    private final DocumentRepository documentsRepo;
    private final UserRepository usersRepo;

    private final SecureRandom secureRandom = new SecureRandom();

    public BetaSessionsService(AppProperties appProperties,
                               ProjectAccessService projectAccessService,
                               BetaAccessService betaAccessService,
                               BetaSessionRepository sessionsRepo,
                               BetaSessionInviteRepository invitesRepo,
                               BetaSessionParticipantRepository participantsRepo,
                               DocumentRepository documentsRepo,
                               UserRepository usersRepo) {
        this.appProperties = appProperties;
        this.projectAccessService = projectAccessService;
        this.betaAccessService = betaAccessService;
        this.sessionsRepo = sessionsRepo;
        this.invitesRepo = invitesRepo;
        this.participantsRepo = participantsRepo;
        this.documentsRepo = documentsRepo;
        this.usersRepo = usersRepo;
    }

    private String hashToken(String raw) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(raw.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(hash);
        } catch (Exception ex) {
            throw new IllegalStateException("hash failed", ex);
        }
    }

    private String randomToken() {
        byte[] bytes = new byte[32];
        secureRandom.nextBytes(bytes);
        return HexFormat.of().formatHex(bytes);
    }

    public List<BetaSession> listForUser(UUID userId, UUID projectId) {
        if (userId == null) return List.of();
        List<UUID> projectIds = projectAccessService.filterAccessibleProjectIds(userId, projectId);
        if (projectIds.isEmpty()) return List.of();
        return sessionsRepo.findByProjectIdInOrderByCreatedAtDesc(projectIds);
    }

    public BetaSession findOneForUser(UUID userId, UUID id) {
        if (userId == null) return null;
        return betaAccessService.getAccess(userId, id).session();
    }

    @Transactional
    public BetaSession createForUser(UUID userId, Map<String, Object> body) {
        if (userId == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "로그인이 필요합니다.");
        Object projectIdObj = body.get("projectId");
        Object titleObj = body.get("title");
        if (!(projectIdObj instanceof String) || ((String) projectIdObj).isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "projectId가 필요합니다.");
        }
        if (!(titleObj instanceof String) || ((String) titleObj).isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "title이 필요합니다.");
        }
        UUID projectId = UUID.fromString((String) projectIdObj);
        projectAccessService.assertProjectAccess(userId, projectId);

        BetaSession session = new BetaSession();
        session.setProjectId(projectId);
        if (body.get("documentId") instanceof String s && !s.isBlank()) session.setDocumentId(UUID.fromString(s));
        session.setTitle((String) titleObj);
        session.setDescription(body.get("description") instanceof String s ? s : null);
        session.setStatus(body.get("status") instanceof String s ? s : "open");
        session.setCreatedBy(userId);
        return sessionsRepo.save(session);
    }

    @Transactional
    public BetaSession updateForUser(UUID userId, UUID id, Map<String, Object> body) {
        if (userId == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "로그인이 필요합니다.");
        BetaSession session = betaAccessService.assertManageAccess(userId, id);
        if (body.containsKey("title")) session.setTitle((String) body.get("title"));
        if (body.containsKey("description")) session.setDescription((String) body.get("description"));
        if (body.containsKey("status")) session.setStatus((String) body.get("status"));
        return sessionsRepo.save(session);
    }

    @Transactional
    public void removeForUser(UUID userId, UUID id) {
        if (userId == null) return;
        BetaSession session = betaAccessService.assertManageAccess(userId, id);
        sessionsRepo.deleteById(session.getId());
    }

    @Transactional
    public Map<String, Object> createInviteForSession(UUID userId, UUID sessionId, Map<String, Object> body) {
        if (userId == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "로그인이 필요합니다.");
        betaAccessService.assertManageAccess(userId, sessionId);

        String rawToken = randomToken();
        String tokenHash = hashToken(rawToken);
        Integer expiresInHours = body.get("expiresInHours") instanceof Number n ? n.intValue() : null;
        Integer maxUses = body.get("maxUses") instanceof Number n ? n.intValue() : null;

        OffsetDateTime expiresAt = null;
        int hours = expiresInHours == null ? 24 * 7 : expiresInHours;
        if (hours > 0) expiresAt = OffsetDateTime.now().plusHours(hours);

        BetaSessionInvite invite = new BetaSessionInvite();
        invite.setSessionId(sessionId);
        invite.setTokenHash(tokenHash);
        invite.setExpiresAt(expiresAt);
        invite.setMaxUses(maxUses);
        invite.setCreatedBy(userId);
        invite = invitesRepo.save(invite);

        String baseUrl = appProperties.getBaseUrl() == null ? "http://localhost:3100" : appProperties.getBaseUrl();
        String inviteUrl = baseUrl.replaceAll("/$", "") + "/beta-invite?token=" + java.net.URLEncoder.encode(rawToken, java.nio.charset.StandardCharsets.UTF_8);

        return Map.of(
            "success", true,
            "inviteId", invite.getId(),
            "token", rawToken,
            "inviteUrl", inviteUrl,
            "expiresAt", invite.getExpiresAt(),
            "maxUses", invite.getMaxUses()
        );
    }

    public Map<String, Object> getInviteInfo(UUID userId, String token) {
        if (userId == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "로그인이 필요합니다.");
        if (token == null || token.isBlank()) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "token이 필요합니다.");

        String tokenHash = hashToken(token);
        BetaSessionInvite invite = invitesRepo.findByTokenHash(tokenHash)
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "초대장을 찾을 수 없습니다."));

        if (invite.getExpiresAt() != null && invite.getExpiresAt().isBefore(OffsetDateTime.now())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "초대장이 만료되었습니다.");
        }
        if (invite.getMaxUses() != null && invite.getUses() >= invite.getMaxUses()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "초대장 사용 횟수를 초과했습니다.");
        }

        BetaSession session = sessionsRepo.findById(invite.getSessionId())
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "베타 세션을 찾을 수 없습니다."));

        return Map.of(
            "session", Map.of(
                "id", session.getId(),
                "projectId", session.getProjectId(),
                "documentId", session.getDocumentId(),
                "title", session.getTitle(),
                "description", session.getDescription(),
                "status", session.getStatus()
            ),
            "invite", Map.of(
                "id", invite.getId(),
                "expiresAt", invite.getExpiresAt(),
                "maxUses", invite.getMaxUses(),
                "uses", invite.getUses()
            )
        );
    }

    @Transactional
    public Map<String, Object> joinByInvite(UUID userId, String token) {
        if (userId == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "로그인이 필요합니다.");
        if (token == null || token.isBlank()) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "token이 필요합니다.");

        String tokenHash = hashToken(token);
        BetaSessionInvite invite = invitesRepo.findByTokenHash(tokenHash)
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "초대장을 찾을 수 없습니다."));

        if (invite.getExpiresAt() != null && invite.getExpiresAt().isBefore(OffsetDateTime.now())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "초대장이 만료되었습니다.");
        }
        if (invite.getMaxUses() != null && invite.getUses() >= invite.getMaxUses()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "초대장 사용 횟수를 초과했습니다.");
        }

        BetaSession session = sessionsRepo.findById(invite.getSessionId())
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "베타 세션을 찾을 수 없습니다."));

        BetaSessionParticipant existing = participantsRepo.findBySessionIdAndUserId(session.getId(), userId).orElse(null);
        if (existing == null) {
            String displayName = usersRepo.findById(userId).map(User::getName).orElse(null);
            BetaSessionParticipant participant = new BetaSessionParticipant();
            participant.setSessionId(session.getId());
            participant.setUserId(userId);
            participant.setStatus("joined");
            participant.setDisplayName(displayName);
            participantsRepo.save(participant);

            invite.setUses(invite.getUses() + 1);
            invitesRepo.save(invite);
        }

        return Map.of("success", true, "sessionId", session.getId());
    }

    public List<Map<String, Object>> listParticipantsForSession(UUID userId, UUID sessionId) {
        if (userId == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "로그인이 필요합니다.");
        BetaSession session = betaAccessService.assertManageAccess(userId, sessionId);

        List<BetaSessionParticipant> participants = participantsRepo.findBySessionIdOrderByJoinedAtAsc(session.getId());
        if (participants.isEmpty()) return List.of();

        List<UUID> userIds = participants.stream().map(BetaSessionParticipant::getUserId).toList();
        Map<UUID, User> userMap = usersRepo.findAllById(userIds).stream().collect(java.util.stream.Collectors.toMap(User::getId, u -> u));

        return participants.stream().map(p -> Map.of(
            "id", p.getId(),
            "sessionId", p.getSessionId(),
            "userId", p.getUserId(),
            "status", p.getStatus(),
            "displayName", p.getDisplayName(),
            "joinedAt", p.getJoinedAt(),
            "user", userMap.getOrDefault(p.getUserId(), null)
        )).toList();
    }

    public Document getSessionDocumentForUser(UUID userId, UUID sessionId) {
        if (userId == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "로그인이 필요합니다.");
        BetaAccessService.AccessResult access = betaAccessService.getAccess(userId, sessionId);
        BetaSession session = access.session();
        if (session.getDocumentId() == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "이 세션에는 documentId가 설정되어 있지 않습니다.");
        }
        Document doc = documentsRepo.findById(session.getDocumentId()).orElse(null);
        if (doc == null) throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "문서를 찾을 수 없습니다.");
        if (!doc.getProjectId().equals(session.getProjectId())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "문서 정보가 올바르지 않습니다.");
        }
        return doc;
    }
}
